README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

## fk_toggle v0.1.4

This is simple flutter toggle button widget.

Supports show text labels and icons, Possible set multiple value to toggle ui, not only 2 values.

Also you can customize size and colors as you like.


## Screenshot of [SampleApp](https://github.com/fukutan-org/fk_toggle/tree/master/example)
![fk_toggle_image](https://user-images.githubusercontent.com/51870919/155354687-a01ddfb5-74d7-49da-80fd-9384fb0bdf32.png)

## Features

- Toggle Items
  - Provide toggle change event.
- Show texts
- Show icons
- Customization of Size, And colors when toggle on/off.

## How to install

Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  fk_toggle: ^0.1.1
```

You can install packages from the command line with Flutter:
```shell
$ flutter pub get
```

Import in your project:
```dart
import 'package:fk_toggle/fk_toggle.dart';
```

## Usage

```dart
Scaffold(
  appBar: AppBar(title: const Text('Sample Toggle Switches')),
  body: Center(
    child: Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: Padding(
          padding: const EdgeInsets.all(10.0),
          child: FkToggle(
            width: 50,
            height: 30,
            labels: const ['1', '2']
          ),
        )
      )
    ),
  ),
);
```

## Parameter descriptions

|Property name|description|default value|required|
|---|---|:-:|:-:|
|width|length of a toogle :warning: not full length of widget|:u7121:|:heavy_check_mark:|
|height|height of widget|:u7121:|:heavy_check_mark:|
|labels|toggle button texts. Allow empty text|:u7121:|:heavy_check_mark:|
|cornerRadius|size of corner radius|10||
|backgroundColor|unselected toggle color|Colors.black26||
|selectedColor|selected toggle color|Colors.blue||
|disabledElementColor|unselected label color|Colors.black54||
|enabledElementColor|selected label color|Colors.white||
|fontSize|font size of labels|16.0||
|icons|Allow null value in lists,You can set an icon for any Toggle|:u7121:||
|onSelected|provide toggle change event, Provide index and instance|:u7121:||

## Future schedule

- Implement animation feature
- Improve class paramters to about like auto calulate width and height.
- you can post some issues.

## Contacts

Feel free to send message.<br>
Welcome to work consultation or everything.<br>
Twitter:[@fukutan_org](https://twitter.com/fukutan_org)<br>
github:[fukutan-org](https://github.com/fukutan-org)


CHANGELOG.md
## [0.1.4] 2022-02-28

Fix readme.

## [0.1.3] 2022-02-27

Improve description in pubspec.yaml that text length 60 to 180 characters.

## [0.1.2] 2022-02-27

Fix readme.md,
Add comments in fk_toggle.

## [0.1.1] 2022-02-27

Fix readme.md.

## [0.1.0] 2022-02-26

Add Example project.

## [0.0.1] 2022-02-25

create library project.


lib/fk_toggle.dart
library fk_toggle;

import 'package:flutter/material.dart';

typedef OnSelected = void Function(int idx, FkToggle instance);

class FkToggle extends StatefulWidget {
  const FkToggle({
    Key? key,
    required this.width,
    required this.height,
    required this.labels,
    this.cornerRadius = 10,
    this.backgroundColor = Colors.black26,
    this.selectedColor = Colors.blue,
    this.disabledElementColor = Colors.black54,
    this.enabledElementColor = Colors.white,
    this.fontSize = 16.0,
    this.icons,
    this.onSelected
  }):   assert(labels.length >= 2),
        assert(width > 0),
        assert(height > 0),
        assert(cornerRadius > 0),
        super(key: key);

  ///
  /// Width of a toggle button length, warning: this is not full length!!<br>
  /// example, width is 50.0 and labels are [1,2] that results of below.<br>
  ///   [  1 (50.0) ][  2 (50.0) ]<br>
  /// So makes a full length 100.0 size toggle buttons include 2 label texts.<br>
  ///
  final double width;

  ///
  /// Height of toggle buttons.
  ///
  final double height;

  ///
  /// Required toggle button texts. Allow empty text.<br>
  /// FkToggle(<br>
  ///   <t>width: 50,<br>
  ///   height: 30,<br>
  ///   labels: const ['1', '2', '', '', '3', '4']<br>
  /// )<br>
  ///<br>
  ///   Result [1][2][][][3][4], So Some toggles has empty text.<br>
  ///   For example case of wants to set icons in empty label buttons.<br>
  ///
  final List<String> labels;

  ///
  /// Typical 10 to 20, Recommended 5 to 30.
  ///
  final double cornerRadius;

  ///
  /// unselected toggle color
  ///
  final Color backgroundColor;

  ///
  /// selected toggle color
  ///
  final Color selectedColor;

  ///
  /// unselected label color
  ///
  final Color disabledElementColor;

  ///
  /// selected label color
  ///
  final Color enabledElementColor;

  ///
  /// font size of labels
  ///
  final double fontSize;

  ///
  /// Allow null value in lists,You can set an icon for any Toggl
  ///
  final List<Icon?>? icons;

  ///
  /// provide toggle change event, Provide index and instance
  ///
  final OnSelected? onSelected;

  @override
  State<StatefulWidget> createState() => _State();
}

class _State extends State<FkToggle> {
  late final Widget _selectedSwitch;
  late final BorderRadius _roundedRadius;

  double _leftMargin = 0.0;
  int _selectedIndex = 0;

  @override
  void initState() {
    super.initState();
    _roundedRadius = BorderRadius.all(Radius.circular(widget.cornerRadius));
    _selectedSwitch = Material(
      color: widget.selectedColor,
      borderRadius: _roundedRadius,
      child: InkWell(
        onTap: () {},
        borderRadius: _roundedRadius,
        child: SizedBox(
          width: widget.width,
          height: widget.height,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      borderRadius: _roundedRadius,
      color: widget.backgroundColor,
      child: InkWell(
        highlightColor: Colors.white54,
        borderRadius: _roundedRadius,
        radius: 50,
        onTap: () {},
        onTapDown: (TapDownDetails details) {
          final int index = details.localPosition.dx ~/ widget.width;
          setState(() {
            _selectedIndex = index;
            _leftMargin = (index * (widget.width));
          });
          if (widget.onSelected != null) {
            widget.onSelected!(index, widget);
          }
          debugPrint('tapDown');
        },
        child: SizedBox(
          width: widget.width * widget.labels.length,
          height: widget.height,
          child: Stack(
            children: [
              Padding(
                padding: EdgeInsets.only(left: _leftMargin),
                child: _selectedSwitch,
              ),
              Positioned.fill(
                child: Align(
                  alignment: Alignment.center,
                  child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: _buildLabels()
                  ),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }

  List<Widget> _buildLabels() {
    final List<Widget> list = [];
    widget.labels.asMap().forEach((idx, label) {
      final child = Expanded(
        flex: 1,
        child: RichText(
          textAlign: TextAlign.center,
          text: TextSpan(
            style: TextStyle(
                fontSize: widget.fontSize,
                color: _selectedIndex == idx
                    ? widget.enabledElementColor
                    : widget.disabledElementColor),
            children: [
              TextSpan(text: label),
              WidgetSpan(child: _tryGet(widget.icons, idx) ?? const SizedBox())
            ]
          ),
        )
      );
      list.add(child);
    });
    return list;
  }

  Icon? _tryGet(List<Icon?>? icons, int idx) {
    if (icons != null && idx < icons.length) {
      return icons[idx];
    }
    return null;
  }
}


example/lib/sample.dart
import 'package:fk_toggle/fk_toggle.dart';
import 'package:flutter/material.dart';

class FkToggleSample extends StatelessWidget {
  const FkToggleSample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final OnSelected selected = ((index, instance) {
      ScaffoldMessenger.of(context).hideCurrentSnackBar();
      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
          content: Text('Select $index, toggle ${instance.labels[index]}')));
    });

    final toggles = [
      FkToggle(
          width: 50,
          height: 30,
          labels: const ['1', '2'],
          onSelected: selected
      ),
      FkToggle(
          width: 100,
          height: 50,
          labels: const ['one', 'two'],
          fontSize: 24,
          cornerRadius: 4,
          onSelected: selected),
      FkToggle(
          width: 50,
          height: 30,
          labels: const ['1', '2', '3', '4', '5'],
          onSelected: selected),
      FkToggle(
          width: 60,
          height: 50,
          labels: const ['ON', 'OFF'],
          cornerRadius: 30,
          onSelected: selected),
      FkToggle(
          width: 60,
          height: 40,
          icons: const [
            Icon(Icons.add_reaction),
            Icon(Icons.add_reaction_outlined)
          ],
          labels: const ['', ''],
          onSelected: selected),
      FkToggle(
          width: 120,
          height: 50,
          labels: const ['Orange', 'White'],
          selectedColor: Colors.orange,
          backgroundColor: Colors.white,
          onSelected: selected),
      FkToggle(
          width: 100,
          height: 50,
          icons: const [
            Icon(Icons.add_reaction),
          ],
          labels: const ['', 'smile'],
          onSelected: selected),
      FkToggle(
          width: 120,
          height: 50,
          selectedColor: Colors.red,
          backgroundColor: Colors.white,
          enabledElementColor: Colors.yellow,
          disabledElementColor: Colors.grey,
          icons: const [
            Icon(Icons.sentiment_satisfied_alt),
            Icon(Icons.emoji_emotions),
          ],
          labels: const ['Smile', 'Good'],
          onSelected: selected),
    ];

    return Scaffold(
      appBar: AppBar(title: const Text('Sample Toggle Switches')),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: toggles
              .map((e) => Padding(
                padding: const EdgeInsets.all(10.0),
                child: e,
              ))
              .toList(),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:fk_toggle/fk_toggle.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        scaffoldBackgroundColor: const Color(0xfff5f5f5),
        primarySwatch: Colors.blue,
      ),
      home: const FkToggleSample(),
    );
  }
}

class FkToggleSample extends StatelessWidget {
  const FkToggleSample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final OnSelected selected = ((index, instance) {
      ScaffoldMessenger.of(context).hideCurrentSnackBar();
      ScaffoldMessenger.of(context).showSnackBar(SnackBar(
          content: Text('Select $index, toggle ${instance.labels[index]}')));
    });

    final toggles = [
      FkToggle(
          width: 50,
          height: 30,
          labels: const ['1', '2'],
          onSelected: selected
      ),
      FkToggle(
          width: 100,
          height: 50,
          labels: const ['one', 'two'],
          fontSize: 24,
          cornerRadius: 4,
          onSelected: selected),
      FkToggle(
          width: 50,
          height: 30,
          labels: const ['1', '2', '3', '4', '5'],
          onSelected: selected),
      FkToggle(
          width: 60,
          height: 50,
          labels: const ['ON', 'OFF'],
          cornerRadius: 30,
          onSelected: selected),
      FkToggle(
          width: 60,
          height: 40,
          icons: const [
            Icon(Icons.add_reaction),
            Icon(Icons.add_reaction_outlined)
          ],
          labels: const ['', ''],
          onSelected: selected),
      FkToggle(
          width: 120,
          height: 50,
          labels: const ['Orange', 'White'],
          selectedColor: Colors.orange,
          backgroundColor: Colors.white,
          onSelected: selected),
      FkToggle(
          width: 100,
          height: 50,
          icons: const [
            Icon(Icons.add_reaction),
          ],
          labels: const ['', 'smile'],
          onSelected: selected),
      FkToggle(
          width: 120,
          height: 50,
          selectedColor: Colors.red,
          backgroundColor: Colors.white,
          enabledElementColor: Colors.yellow,
          disabledElementColor: Colors.grey,
          icons: const [
            Icon(Icons.sentiment_satisfied_alt),
            Icon(Icons.emoji_emotions),
          ],
          labels: const ['Smile', 'Good'],
          onSelected: selected),
    ];

    return Scaffold(
      appBar: AppBar(title: const Text('Sample Toggle Switches')),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: toggles
              .map((e) => Padding(
            padding: const EdgeInsets.all(10.0),
            child: e,
          ))
              .toList(),
        ),
      ),
    );
  }
}
